src/Microsoft.OData.Core/UriParser/CustomUriFunctions.cs

```diff
@@ -22,196 +21,228 @@ namespace Microsoft.OData.UriParser
                         }
                     }
                 }
 
-                // Setup the output values.
-                nameSignatures = bufferedKeyValuePairs.Count != 0 ? bufferedKeyValuePairs : null;
-
-                return nameSignatures != null;
+                return functionSignatures != null && functionSignatures.Count > 0;
             }
         }
 
         #endregion
 
-        #region Private Methods
+        #region Internal Methods
 
-        private static void AddCustomFunction(string customFunctionName, FunctionSignatureWithReturnType newCustomFunctionSignature)
+        /// <summary>
+        /// Retrieves the <see cref="CustomUriFunctionsStore"/> instance associated with the given Edm model,
+        /// creating and annotating it on the model if it does not already exist.
+        /// </summary>
+        /// <param name="model">The Edm model for which to get or create the custom URI functions store.</param>
+        /// <returns>The <see cref="CustomUriFunctionsStore"/> instance for the model.</returns>
+        internal static CustomUriFunctionsStore GetOrCreateStore(this IEdmModel model)
         {
-            FunctionSignatureWithReturnType[] existingCustomFunctionOverloads;
-
-            // In case the function doesn't already exist
-            if (!CustomFunctions.TryGetValue(customFunctionName, out existingCustomFunctionOverloads))
-            {
-                CustomFunctions.Add(customFunctionName, new FunctionSignatureWithReturnType[] { newCustomFunctionSignature });
-            }
-            else
-            {
-                // Function does already exist as a custom function in cache
-                // Check if the function overload doesn't already exist
-                bool isOverloadAlreadyExist =
-                    existingCustomFunctionOverloads.Any(existingFunction => AreFunctionsSignatureEqual(existingFunction, newCustomFunctionSignature));
+            return CustomUriFunctionsStore.GetOrCreate(model);
+        }
 
-                if (isOverloadAlreadyExist)
-                {
-                    // Throw if already exists - User is stupid (inserted the same function twice)
-                    throw new ODataException(Error.Format(SRResources.CustomUriFunctions_AddCustomUriFunction_CustomFunctionOverloadExists, customFunctionName));
-                }
+        /// <summary>
+        /// Attempts to retrieve the <see cref="CustomUriFunctionsStore"/> instance associated with the given Edm model,
+        /// or returns <c>null</c> if no store has been created or annotated on the model.
+        /// </summary>
+        /// <param name="model">The Edm model for which to retrieve the custom URI functions store.</param>
+        /// <returns>
+        /// The <see cref="CustomUriFunctionsStore"/> instance if it exists on the model; otherwise, <c>null</c>.
+        /// </returns>
+        internal static CustomUriFunctionsStore TryGetStore(this IEdmModel model)

```

Comment:
nit: When called as `model.TryGetStore()` it gives the impression that it's a general-purpose store.

```suggestion
        internal static CustomUriFunctionsStore TryGetCustomUriFunctionStore(this IEdmModel model)

```

Reply:
For the same reasoning I have given here https://github.com/OData/odata.net/pull/3322#discussion_r2275569624, I'll adopt `TryGet` here

Reply:
UPDATE: I'll drop this extension method and have the methods call `CustomUriFunctionStore.TryGet`